supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala

```diff
@@ -0,0 +1,137 @@
+ * best way to do it is during Cromwell startup.
+ *
+ */
+case class PipelinesApiReferenceFilesMapping(validReferenceFilesMap: Map[String, PipelinesApiReferenceFilesDisk])
+
+object PipelinesApiReferenceFilesMapping extends PipelinesApiReferenceFilesMappingOperations
+
+// functionality extracted into the trait for testing purposes
+protected trait PipelinesApiReferenceFilesMappingOperations {
+  private val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  case class ReferenceFile(path: String, crc32c: Long)
+  case class ManifestFile(imageIdentifier: String, diskSizeGb: Int, files: List[ReferenceFile])
+
+  def generateReferenceFilesMapping(auth: GoogleAuthMode,
+                                    referenceDiskLocalizationManifestFiles: Option[List[ValidFullGcsPath]]): PipelinesApiReferenceFilesMapping = {
+
+    val validReferenceFilesMapIO = referenceDiskLocalizationManifestFiles match {
+      case None =>
+        IO.pure(Map.empty[String, PipelinesApiReferenceFilesDisk])
+      case Some(manifestFilesPaths) =>
+        val gcsClient = StorageOptions
+          .newBuilder()
+          .setCredentials(auth.credentials(Set(StorageScopes.DEVSTORAGE_READ_ONLY)))
+          .build
+          .getService
+        val manifestFilesIo = manifestFilesPaths traverse { manifestFilePath => readReferenceDiskManifestFileFromGCS(gcsClient, manifestFilePath) }
+        manifestFilesIo flatMap { manifestFiles =>
+          manifestFiles
+            .traverse(manifestFile => getMapOfValidReferenceFilePathsToDisks(gcsClient, manifestFile))
+            .map(_.flatten.toMap)
+        }
+    }
+    PipelinesApiReferenceFilesMapping(validReferenceFilesMapIO.unsafeRunSync())
+  }
+
+  def getReferenceDisksToMount(pipelinesApiReferenceFilesMapping: PipelinesApiReferenceFilesMapping,
+                               inputFilePaths: Set[String]): List[PipelinesApiReferenceFilesDisk] = {
+    pipelinesApiReferenceFilesMapping.validReferenceFilesMap.filterKeys(key => inputFilePaths.contains(s"gs://$key")).values.toList.distinct
+  }
+
+  protected def readReferenceDiskManifestFileFromGCS(gcsClient: Storage, gcsPath: ValidFullGcsPath): IO[ManifestFile] = {
+    val manifestFileBlobIo = IO { gcsClient.get(BlobId.of(gcsPath.bucket, gcsPath.path.substring(1))) }
+    manifestFileBlobIo flatMap { manifestFileBlob =>
+      val jsonStringIo = IO { manifestFileBlob.getContent().map(_.toChar).mkString }
+      jsonStringIo.flatMap(jsonStr => IO.fromEither(decode[ManifestFile](jsonStr)))
+    }
+  }
+
+  private def getReferenceFileToValidatedGcsPathMap(referenceFiles: Set[ReferenceFile]): IO[Map[ReferenceFile, ValidFullGcsPath]] = {
+    val filesAndValidatedPaths = referenceFiles.map {
+      referenceFile => (referenceFile, GcsPathBuilder.validateGcsPath(s"gs://${referenceFile.path}"))
+    }.toMap
+
+    val (filesWithValidPaths, filesWithInvalidPaths) = filesAndValidatedPaths.partition(_._2.isInstanceOf[ValidFullGcsPath])
```

Comment:
I actually almost preferred the dual `.collect`s version you had before because it guaranteed the types for you and avoided the awkward `_.asInstanceOf`s. 

I wonder whether there's an equivalent that does a single pass over a list but divides into two well-typed outputs...

UPDATE: based on quick stackoverflow searching, this seems like the closest thing: https://stackoverflow.com/a/24216337

Reply:
I don't like the `foldRight` approach since it's not immediately obvious what's happening there, so every time while reading the code I'd have to stop and think.